segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ConcurrencyManager.java

```diff
@@ -0,0 +1,247 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.impl.bookkeeper;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import io.pravega.common.AbstractTimer;
+import io.pravega.common.MathHelpers;
+import java.util.function.Supplier;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * Manages the Write Queue Concurrency.
+ */
+@ThreadSafe
+class ConcurrencyManager {
+    //region Members
+
+    @VisibleForTesting
+    static final double SIGNIFICANT_DIFFERENCE = 0.25; // Changes below this will be ignored.
+    @VisibleForTesting
+    static final long UPDATE_PERIOD_MILLIS = 1000; // Min interval for updates.
+    @VisibleForTesting
+    static final long STALE_MILLIS = UPDATE_PERIOD_MILLIS * 15; // Reset all stats if no activity in this interval.
+    @VisibleForTesting
+    static final int MAX_STAGNATION_AGE = 10;
+    private final int minParallelism;
+    private final int maxParallelism;
+    private final Supplier<Long> timeSupplier;
+    private final Object snapshotLock = new Object();
+    @GuardedBy("this")
```

Comment:
Given that `this` is guarding private variables, should we consider locking them on a private object rather than `this`?